addons/stock/models/stock_move.py

```diff
@@ -1235,32 +1231,96 @@ def _should_bypass_reservation(self):
         return self.location_id.should_bypass_reservation() or self.product_id.type != 'product'
 
     # necessary hook to be able to override move reservation to a restrict lot, owner, pack, location...
-    def _get_available_quantity(self, location_id, lot_id=None, package_id=None, owner_id=None, strict=False, allow_negative=False):
+    def _get_available_quantity(self, location_id, lot_id=None, package_id=None, owner_id=None, strict=False, allow_negative=False, quants=None):
         self.ensure_one()
-        return self.env['stock.quant']._get_available_quantity(self.product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict, allow_negative=allow_negative)
+        return self.env['stock.quant']._get_available_quantity(self.product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict, allow_negative=allow_negative, quants=quants)
 
     def _action_assign(self):
         """ Reserve stock moves by creating their stock move lines. A stock move is
         considered reserved once the sum of `product_qty` for all its move lines is
         equal to its `product_qty`. If it is less, the stock move is considered
         partially available.
         """
-        assigned_moves = self.env['stock.move']
-        partially_available_moves = self.env['stock.move']
-        # Read the `reserved_availability` field of the moves out of the loop to prevent unwanted
+        moves_to_assign = self.filtered(lambda m: m.state in ['confirmed', 'waiting', 'partially_available'])
+
+        # Cache moves which should bypass
+        moves_to_bypass = moves_to_assign.filtered(lambda move: move._should_bypass_reservation())
+        moves_to_bypass_ids = set(moves_to_bypass.ids)
+
+        # Cache the empty move
+        def is_empty_move(move):
+            return float_is_zero(move.product_uom_qty, precision_rounding=move.product_uom.rounding)
```

Comment:
can be replaced by `fields.Float.is_zero`
https://github.com/odoo/odoo/blob/14.0/odoo/fields.py#L1290